subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,1023 @@
+ * is a part of the device. If so then true is returned and in parameter
+ * audio_dev_data is considered correct device data to be manipulated by the
+ * request.
+ *
+ * @note For now this function searches through feature units only. The
+ *	 descriptors are known and are not using any other entity type.
+ *	 If there is a need to add other units to audio function then this
+ *	 must be reworked.
+ *
+ * @param [in]      audio_dev_data USB audio device data.
+ * @param [in, out] entity	   USB Audio entity addressed by the request.
+ *				   .id      [in]  id of searched entity
+ *				   .subtype [out] subtype of entity (if found)
+ *
+ * @return true if entity matched audio_dev_data, false otherwise.
+ */
+static bool is_entity_valid(struct usb_audio_dev_data_t *audio_dev_data,
+			    struct usb_audio_entity *entity)
+{
+	const struct cs_ac_interface_descriptor_header *header;
+	struct feature_unit_descriptor *fu;
+
+	header = audio_dev_data->header_descr;
+	fu = (struct feature_unit_descriptor *)((u8_t *)header +
+						header->bLength +
+						INPUT_TERMINAL_DESC_SIZE);
+	if (fu->bUnitID == entity->id) {
+		entity->subtype = fu->bDescriptorSubtype;
+		return true;
+	}
+
+	if (header->bInCollection == 2) {
+		fu = (struct feature_unit_descriptor *)((u8_t *)fu +
+			fu->bLength +
+			INPUT_TERMINAL_DESC_SIZE +
+			OUTPUT_TERMINAL_DESC_SIZE);
+		if (fu->bUnitID == entity->id) {
+			entity->subtype = fu->bDescriptorSubtype;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+struct usb_audio_dev_data_t *get_audio_dev_data_by_entity(struct usb_audio_entity *entity)
+{
+	struct usb_dev_data *dev_data;
+	struct usb_audio_dev_data_t *audio_dev_data;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&usb_audio_data_devlist, dev_data, node) {
+		LOG_DBG("iterate over dev_data in audio list");
+		audio_dev_data = CONTAINER_OF(dev_data, struct usb_audio_dev_data_t, common);
+
+		if (is_entity_valid(audio_dev_data, entity)) {
+			return audio_dev_data;
+		}
+	}
+	return NULL;
+}
+
+/**
+ * @brief Handler for feature unit requests.
+ *
+ * This function handles feature unit specific requests.
+ * If request is properly served 0 is returned. Negative errno
+ * is returned in case of an error. This leads to setting stall on IN EP0.
+ *
+ * @param [in]  dev_data USB audio device data.
+ * @param [in]  pSetup   Information about the executed request.
+ * @param [in]  len      Size of the buffer.
+ * @param [out] data     Buffer containing the request result.
+ *
+ * @return 0 if succesfulf, negative errno otherwise.
+ */
+static int handle_feature_unit_req(struct usb_audio_dev_data_t *audio_dev_data,
+				   struct usb_setup_packet *pSetup,
+				   s32_t *len,
+				   u8_t **data)
+{
+	enum usb_audio_feature_unit_control_selectors control_selector;
+	struct feature_unit_descriptor *fu = NULL;
+	struct usb_audio_feature_unit_evt evt = { .dev = audio_dev_data->common.dev };
+	u8_t ch_num, ch_start, ch_end;
+	static u8_t tmp_data_ptr[3];
```

Comment:
This assumes that there at most 3 channels, and each control uses an `u8_t` value. Is that guaranteed some way?

Reply:
It's not. I forgot about that. Will think about rework.